Lower quant/dequant torch op to StableHLO

lsy323 opened the pull request:
The following [torch ops](https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py) can be lowered to StableHLO with this diff:
- quantize_per_tensor
- quantize_per_channel
- dequantize_per_tensor
- dequantize_per_channel

#### User Experience
- The GraphModule generated from PT2E quantization can be exported to StableHLO, or tf.saved_model using the existing exporting API without any additional change on model code, or exporting script. `STABLEHLO_BYTECODE_FROM_PRETTYPRINT` needs to be set to 1 to workaround a StableHLO bytecode serialization [issue](https://github.com/openxla/stablehlo/issues/1812).

#### Current workflow
1. Register xla qdq ops to 'XLA' dispatch key. So the qdq ops will be dispatched to xla impl during LTC tracing.
2. During lowering, qdq ops are lowered to a custom call to `stablehlo.uniform_quantize/dequantize` in HLO. The qparams are stored in the custom call config str. The config str can be deserialized to mlir DictAttr directly.
3. HLO->StableHLO converter will convert custom call to `stablehlo.uniform_quantize/dequantize`

#### Changes
1. Allow `save_torch_module_as_tf_saved_model` to take GraphModule as well, since PT2E outputs a GraphModule.
2. Added 2 patches. One is to add support to HLO->StableHLO converter for `stablehlo.uniform_quantize/dequantize` conversion, originally authored by @sdasgup3. Another is to workaround a StableHLO bytecode serialization issue mentioned above. Both won't be needed if HLO qdtype representation is added.
3. Added new xla quantize_tensor/dequantize_tensor ops for qdq ops lowering. the xla quantize/dequantize op lowers to custom call to `stablehlo.uniform_quantize/dequantize`
4. Test script including exporting per-tensor/channel qdq ops and PT2E quantized resnet18 model.

#### Future Work
1. When qdtype is added to HLO, the lowering logic need to be updated and will be more concise than the current one.

cc @sdasgup3 @GleasonK @paulinesho

JackCaoG commented on WORKSPACE:
```diff
@@ -43,6 +43,8 @@ http_archive(
         "//openxla_patches:gpu_race_condition.diff",
         "//openxla_patches:f16_abi_clang.diff",
         "//openxla_patches:gpu_topk_rewriter.diff",
+        "//openxla_patches:quant_dequant_converter.diff",
+        "//openxla_patches:stablehlo_quant_seralization.diff",
```
are these changes already landed on openxla?

lsy323 replied:
No, temporary patch before qdtype in HLO is landed

lsy323 commented on torch_xla/tf_saved_model_integration.py:
```diff
@@ -120,7 +121,7 @@ def save_stablehlo_graph_as_tf(
 
 def save_torch_module_as_tf_saved_model(
-    torch_model: torch.nn.Module,
+    torch_model: Union[torch.nn.Module, GraphModule],
```
Yeah, that's right, let me keep it as it is.

lsy323 commented on torch_xla/csrc/runtime/stablehlo_helper.cc:
```diff
@@ -37,13 +37,23 @@ static std::string getMlirModuleStr(mlir::ModuleOp& mlir_module) {
 
-static std::string getMlirModuleBytecode(const mlir::ModuleOp& mlir_module) {
+static std::string getMlirModuleBytecode(mlir::ModuleOp& mlir_module) {
+  static bool from_pretty_print = runtime::sys_util::GetEnvBool(
```
This is to workaround a StableHLO bytecode serialization issue at https://github.com/openxla/stablehlo/issues/1812. If the bytecode is serialized from in-memory `mlir:ModuleOp`, and there are `stablehlo.uniform_quantize/dequantize` in the StableHLO graph, the bytecode serialization would fail, however, txt format serialization can success. The workaround here is to serialized `mlir::ModuleOp` to txt format first, then deserialize to `mlir::ModuleOp`, then serialized to bytecode format.

lsy323 commented on torch_xla/csrc/quant_util.h:
```diff
@@ -0,0 +1,41 @@
+
+namespace torch_xla {
+
+struct QuantParams {
```
Will add.

lsy323 commented on torch_xla/csrc/ops/dequant_tensor.h:
```diff
@@ -0,0 +1,32 @@
+
+}  // namespace torch_xla
+
+#endif  // XLA_TORCH_XLA_CSRC_OPS_QUANT_TENSOR_H_
```
Done

lsy323 commented on torch_xla/csrc/init_python_bindings.cpp:
```diff
@@ -1104,6 +1126,30 @@ void InitXlaModuleBindings(py::module m) {
+           int quant_max, const std::string& dtype, int axis) -> at::Tensor {
+          at::Tensor result;
+          {
+            NoGilSection nogil;
```
Synced offline, keep it as all current custom ops release the GIL during tracing.

lsy323 commented on test/stablehlo/test_pt2e_qdq.py:
```diff
@@ -0,0 +1,72 @@
+    exported = torch.export.export(m, args)
+    stablehlo_gm = stablehlo.exported_program_to_stablehlo(exported)
+    stablehlo_txt = stablehlo_gm.get_stablehlo_text('forward')
+    # print(stablehlo_txt)
```
ok to remove

lsy323 commented on test/stablehlo/test_pt2e_qdq.py:
```diff
@@ -0,0 +1,72 @@
+    self.assertTrue("stablehlo.uniform_quantize" in stablehlo_txt)
+    self.assertTrue("stablehlo.uniform_dequantize" in stablehlo_txt)
+    # Save as tf.saved_model
+    # save_torch_module_as_tf_saved_model(m, args, '/tmp/tf_saved_model/tmp1')
```
Intend to keep it there, but checking if the exported tf.saved model is as expected cannot be done easily, unless write a script to load the saved model.

lsy323 replied:
Let me check the existence of the `.pb` file, which should be the main artifact of the tf savedmodel

lsy323 commented on test/stablehlo/test_pt2e_qdq.py:
```diff
@@ -0,0 +1,72 @@
+    stablehlo_gm = stablehlo.exported_program_to_stablehlo(exported)
+    stablehlo_txt = stablehlo_gm.get_stablehlo_text('forward')
+    # print(stablehlo_txt)
+    self.assertTrue("stablehlo.uniform_quantize" in stablehlo_txt)
```
I can assert the specific ops, maybe a better way is to check the op in in-memory format, but that requires introducing mlir python API, which is not trivial.

lsy323 commented on test/stablehlo/test_pt2e_qdq.py:
```diff
@@ -0,0 +1,72 @@
+    # Step 1: export resnet18
+    args = (torch.randn(1, 3, 224, 224),)
+    m = torchvision.models.resnet18().eval()
+    m = capture_pre_autograd_graph(m, args)
```
Here the graph is captured for PT2E to further process. PT2E doesn't work graph captured from torch.exported (just tried locally), it needs to capture the graph in this way.

lsy323 commented:
Update:
- Addressed review comments
- Enhanced testing script to check the qparam of qdq stablehlo ops, numbers of qdq ops
- Added more assertions to the torch_xla qdq ops, including scale, zero_point shape, zero_point dtype matches int dtype of quantized type, scale values are all positive

miladm commented on test/stablehlo/test_pt2e_qdq.py:
```diff
@@ -0,0 +1,117 @@
+    save_torch_module_as_tf_saved_model
+
+# Needed to workaround the stablehlo bytecode serialization issue in https://github.com/openxla/stablehlo/issues/1812
+os.environ['STABLEHLO_BYTECODE_FROM_PRETTYPRINT'] = '1'
```
@lsy323 I assume this is a temp flag we will remove to improve usability quality in near future?

lsy323 replied:
This is to workaround a stablehlo issue. Resolved in https://github.com/pytorch/xla/pull/6613
